src/basics.ml

```diff
@@ -34,6 +35,18 @@ let to_tvar : term -> tvar = fun t ->
     “marshaled” (e.g., by the {!module:Sign} module), as this would break the
     freshness invariant of new variables. *)
 
+(** [stamp_tvar s v] creates a fresh variable whose name is the name of [v]
+    with "s[s]" as suffix.  For instance [stamp_tvar 13 x] with [x] having
+    name [c] will result in a variable with name [cs13]. *)
+let stamp_tvar : int -> tvar -> tvar = fun stamp v ->
+  let name = Bindlib.name_of v ^ "s" ^ (string_of_int stamp) in
+  Bindlib.new_var mkfree name
+
+(** [sensible_tref t] transforms {!constructor:Appl} into references. *)
+let sensible_tref : term -> term = function
+  | Appl(_,_) as t -> TRef(ref (Some t))
```

Comment:
Why this matching? Why not converting any term to TRef and call this function to_tref for instance?

Reply:
We want to avoid creating too many references for performance reasons

Reply:
I see. But why do you need a reference at all? In Rodolphe's implem, it was used for not normalizing an argument twice. Don't decision trees ensure this property by construction?

Reply:
Here the references are used to have sharing: even if the matching fails, reductions performed are not discarded

Reply:
But what you said is true.

Reply:
I see. But it would be good to add a comment somewhere about that. Note that, originally, TRef was intented to be used for the implem of the rewrite tactic only. Note also that, in your case, you cannot have TRef(ref None) I think. But it does not matter much.

Reply:
That's true that I can't, or at least I shouldn't have a `TRef(ref None)`.